### PR TITLE
Fix incorrect high contrast mode detection

### LINKED ISSUE #4379: A11y: Buttons present inside editor section are not visible in High Contrast Black mode.
Any environment

When using high contrast black mode, some buttons are not visible due to Opacity property of ckeditor button class. (Currently 0.3 for disabled buttons)

image

​OS Details:​
Windows Server 2019 Version 1809(OS Build 17763.914)
​
Browser Details:​
Chromium Edge Version 81.0.416.58 (Official build) beta (64-bit)

​Repro Steps:​​
1.Turn on High Contrast and open any ckeditor instance, demo site would suffice.

7.Verify user is able to view the buttons inside editor section or not.
​​
Actual Result:​​
Buttons present inside editor section are not visible in High Contrast Black mode.
Expected Result:​​
Buttons present inside editor section should be visible in High Contrast Black mode.
User Impact:​​
Low vision users can not see all controls if few controls are not visible in high contrast mode.
​
MAS Reference:​​
https://microsoft.sharepoint.com/:w:/r/sites/accessibility/_layouts/15/Doc.aspx?sourcedoc=%7B0BF582E1-6800-4E43-8947-DA015AA28B2B%7D&file=MAS%201.4.3%20%E2%80%93%20Contrast%20(Minimum).docx&action=default&mobileredirect=true

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4379](https://github.com/ckeditor/ckeditor4/issues/4379): [Edge] Fixed: incorrect detection of the high contrast mode.
```

## What changes did you make?

I've enabled detecting HC also in Webkit/Blink. Additionally added the manual test for https://dev.ckeditor.com/ticket/5429

I didn't create unit tests as HC mode is really hard to detect programatically.

## Which issues does your PR resolve?

Closes #4379.

### CODE REVIEW

File: tests/core/ckeditor/manual/hcdetection.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+@bender-tags: feature, 4.15.2, 4379
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, undo, clipboard, basicstyles, about, stylescombo, format, link, image
+
+1. Enable high contrast mode in your OS.
```
Comment (f1ames) [inline tests/core/ckeditor/manual/hcdetection.md:5]: This is a little inconsistent with expected - this step assumes one can enable HC mode and Expected/Unexpected mentions cases where HC mode is not supported by OS which makes it somewhat confusing.

At least we need better description or split this into two parts (OS supporting HC mode and the ones without it).

**Review by f1ames — CHANGES_REQUESTED** [review]
Ok, so I have tested it with HC mode on various browsers and OSes.

## Windows 10

### Edge, IE11, Firefox

Setting HC mode on Windows 10 works fine with Edge, IE11 and Firefox and it is correctly detected and handled by CKEditor 4.

### Chrome

Chrome doesn't really support HC mode out-of-the-box and recommends [HC mode extension](https://chrome.google.com/webstore/detail/high-contrast/djcfdncoelnlbldjfhinnjlhdjlikmph?hl=en), using Dark mode (with some dark themes) or going with experimental features.

With HC mode extension, the contrast is similar to native support but CKEditor 4 doesn't recognize it correctly (probably due to our detection mechanism).

Dark mode doesn't really change the way pages are displayed as long as they don't support it - so there is no effect for CKEditor 4 itself too.

The 3rd option is to use experimental feature so go to `chrome://flags` and enable `Forced colors` - this turns on HC mode support and works correctly the same wa as in Edge.

**I would say this is fine. We can think of extracting support for HC mode extension and dark mode as feature requests** :thinking:

## Windows 7

### IE8

Still works fine :+1: 

## Linux

On Linux (Mint) HC mode doesn't really change the way both Chrome and Firefox renders content (which probably means lack of support), but that's a separate issue we can look later on if needed.

## macOS

There is [Invert Colors option](https://support.apple.com/lv-lv/guide/mac-help/unac089/mac) but it's not exactly the same as Windows HC mode and will not be correctly detected by CKEditor 4.

## Mobiles

As for mobiles, both Android and iOS supports Invert Colors / Text Contrast features (Android [1](https://support.google.com/accessibility/android/answer/6151855?hl=en&ref_topic=9079043) and [2](https://support.google.com/accessibility/android/answer/6151800?hl=en&ref_topic=9079043), [iOS](https://support.apple.com/en-us/HT207025)), but from what I have checked (on Android) they are not detected by our detection mechanism since the styling (border color) may be different. **This is the same case most probably as for macOS and Linux**.

## Summing up

If you take a look at our [High Contrast documentation section](https://ckeditor.com/docs/ckeditor4/latest/guide/skin_sdk_hc.html) it states only Windows HC feature support which is in tact with the above. This PR will require slight adjustments there - to mention at least Edge support (and maybe Chrome experimental one too), @Comandeer could you take care of this too?

So the fix is fine, we only need to polish test slightly so they are not too confusing and are only run in OSes/browsers which are supported. Please see also review comments regarding test.

**Comment by Comandeer** [comment]
Probably we will be able to introduce better support for Chrome and mobile browsers after rewriting the detection mechanism to use [`forced-colors` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors) by default and fall back to the current mechanism in case of older browsers. However, at the moment the support for `forced-colors` is non-existent/experimental.

File: tests/core/ckeditor/manual/hcdetection.md (hunk 1 above)
Comment (Comandeer) [reply tests/core/ckeditor/manual/hcdetection.md:5]: Probably splitting test into two will be the most readable solution.

**Comment by Comandeer** [comment]
Ok, I've updated tests and prepared PR with changes to documentation → ckeditor/ckeditor4-docs#340.

**Comment by f1ames** [comment]
> Probably we will be able to introduce better support for Chrome and mobile browsers after rewriting the detection mechanism to use [`forced-colors` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors) by default and fall back to the current mechanism in case of older browsers. However, at the moment the support for `forced-colors` is non-existent/experimental.

@Comandeer do you think `forced-colors` is also used for mobiles/macOS? Or we will need yet different mechanism to detect it on macOS/iOS/Android? Anyway, @Comandeer could you extract it as a follow-up?

**Review by f1ames — APPROVED** [review]
LGTM :+1: 

---

I was thinking if we shouldn't automatically ignore manual tests based on `CKEDITOR.env.hc`, but then if this detection will not/stop working we won't see it as tests will be auto ignored. So keeping it as is will help in such cases :+1:
